src/pi_cond.c

```diff
@@ -96,6 +96,7 @@ int pi_cond_timedwait(pi_cond_t *cond, pi_mutex_t *mutex,
 				pi_mutex_lock(&cond->priv_mut);
 				cond->pending_wait--;
 				pi_mutex_unlock(&cond->priv_mut);
+				pi_mutex_lock(mutex);
```

Comment:
Again - agree. However, in addition to this issue, the following from [1] is interesting: "Except for [ETIMEDOUT], [ENOTRECOVERABLE], and [EOWNERDEAD], all these error checks shall act as if they were performed immediately at the beginning of processing for the function and shall cause an error return, in effect, prior to modifying the state of the mutex specified by mutex or the condition variable specified by cond." I think we need to carefully consider the first pi_mutex_unlock(mutex) in this routine against this criteria. That's another patch, but wanted to note it.

Reply:
Yeah, I agree. I also think it's going to be tricky with the current futex API since we have to drop the user mutex before doing the FUTEX_WAIT_REQUEUE_PI which in turn can return things like -EAGAIN etc. I'll have to scratch my head a bit longer on this one.
I do have some commits in the queue that fix some spot problems like:
https://github.com/gratian/librtpi/commit/25f376fcf13e68dadcd741e39f937b99b98ad862 ("pi_mutex: Return EINVAL for invalid mutexes")
https://github.com/gratian/librtpi/commit/cd5465797513d7fc170d24f03b5c31d114182089 ("pi_cond: Fix errno getting overwritten ")
https://github.com/gratian/librtpi/commit/1e2f017667dd55c3dfc7e39af22d3e8485ad1457 ("pi_cond: Add check for invalid timeout ")